src/workers/newView.ts

```diff
@@ -56,56 +56,63 @@ RETURNING "currentStreak", "maxStreak"
 const incrementReadingStreak = async (
   con: DataSource,
   data: DeepPartial<View>,
-): Promise<boolean> => {
-  const users = con.getRepository(User);
-  const repo = con.getRepository(UserStreak);
-  const { userId, timestamp } = data;
+): Promise<boolean> =>
+  con.transaction(async (entityManager) => {
+    const users = entityManager.getRepository(User);
+    const repo = entityManager.getRepository(UserStreak);
+    const { userId, timestamp } = data;
 
-  if (!userId) {
-    return false;
-  }
+    if (!userId) {
+      return false;
+    }
 
-  const viewTime = timestamp ? new Date(timestamp as string) : new Date();
+    const viewTime = timestamp ? new Date(timestamp as string) : new Date();
 
-  // TODO: This code is temporary here for now because we didn't run the migration yet,
-  // so not every user is going to have a row in the user_streak table and we need to create it.
-  // We can get rid of it once/if we implement the migration in https://dailydotdev.atlassian.net/browse/MI-70
-  const user = await users.findOne({
-    where: { id: userId },
-    relations: ['streak'],
-  });
-  const streak = await user?.streak;
-  if (user && !streak) {
-    await repo.save(
-      repo.create({
-        userId,
-        currentStreak: 1,
-        totalStreak: 1,
-        maxStreak: 1,
-        lastViewAt: viewTime,
-      }),
-    );
-  } else {
-    const results = await repo.query(INC_STREAK_QUERY, [
+    // TODO: This code is temporary here for now because we didn't run the migration yet,
+    // so not every user is going to have a row in the user_streak table and we need to create it.
+    // We can get rid of it once/if we implement the migration in https://dailydotdev.atlassian.net/browse/MI-70
+    const user = await users.findOne({
+      where: { id: userId },
+      relations: ['streak'],
+    });
+
+    const streak = await user?.streak;
+    // we need to check before we actually update the max streak
+    // otherwise we end up in a spot where we create an alert record
+    // every time a post is viewed, even if we already alerted the user
+    const { currentStreak, maxStreak } = streak || {
+      currentStreak: 0,
+      maxStreak: 0,
+    };
+
+    // milestones are currently defined on fibonacci sequence
+    const showStreakMilestone =
+      isFibonacci(currentStreak + 1) &&
+      streak?.lastViewAt?.getDate() !== viewTime.getDate();
```

Comment:
This is not going to work nice with user timezones. The backend runs on UTC, but the streak calculations run on whatever timezone the user has setup.

Imagine the following scenario:
User is in PST timezone (UTC - 8h). The user reads an article at `2024-03-05 22:12:34-08:00`, which is `2024-03-06T06:12:34Z`. Last view at is `2024-03-05 14:00:52-08:00`, which is `2024-03-05T22:00:52Z`. So `showStreakMilestone` would be true here, even though we are not actually increasing the streak in the database with the `INC_STREAK_QUERY` in  the end.

I think we should instead rely on the previous logic. We can return `shouldIncrementStreak` from the `INC_STREAK_QUERY`. That way we know if the reading streak was updated on this view or not. If not, we don't update the alerts either.

Another possibility would be to modify the `INC_STREAK_QUERY` to also do the alert table modification. That way you can rely on DB logic completely.

Reply:
To put money where my mouth is, created an alternative approach PR: https://github.com/dailydotdev/daily-api/pull/1735